### ISSUE TITLE
PUT is like GET, not POST

### ISSUE THREAD

**Issue by David263** [mdn/content#18614]
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT

### What specific section or headline is this issue about?

2nd paragraph

### What information was incorrect, unhelpful, or incomplete?

The article says, "The difference between PUT and [POST](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/POST) is that PUT is idempotent..."

An experiment done on a local Apache server under Windows 10 with Firefox and with Chrome shows that a PUT form is actually sent as a GET, with the fields explicit in the URL argument list.

It is a big failing that POST hides its arguments but causes an error on resubmission. In other words, whether or not the arguments are hidden should be separated from whether the operation is idempotent. I often wish I could use POST to hide its arguments, but allow the user to resubmit the page just like GET.

### What did you expect to see?

It should give the correct information that PUT is (at least currently) a synonym for GET.

### Do you have any supporting links, references, or citations?

I can't find any documentation that PUT is implemented as GET on the Web. Maybe it's something about my environment? I'm using a full URL in the Action attribute.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/http/methods/put`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/http/methods/put/index.md
* Last commit: https://github.com/mdn/content/commit/0dbdb1c819944887d14ada153bf5ee11390a08e3
* Document last modified: 2022-06-22T05:54:43.000Z

</details>

**Comment by lionralfs** [comment]
It sounds like you're trying to use `method="put"` on a `form`, which is [not supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-method) ([spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-method)), causing the browser to fall back to a `GET`.

**Comment by teoli2003** [comment]
Yes, this is likely the problem. Closing.

**Comment by David263** [comment]
Thank you! Makes sense. So we are stuck with only GET and POST, and POST is stuck with no way to eliminate the resubmission error, enforced by the browsers. Perhaps something similar to that statement could be added to MDN, to save people some confusion.

**Comment by lionralfs** [comment]
I don't know exactly what you're trying to achieve but you could redirect after a `POST` (see [this](https://en.wikipedia.org/wiki/Post/Redirect/Get) for example) or write some JavaScript that captures the form submission and implement custom logic.

**Comment by hamishwillee** [comment]
> It sounds like you're trying to use method="put" on a form, which is [not supported](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-method) ([spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#attr-fs-method)), causing the browser to fall back to a GET.

@lionralfs I'm a little confused too, but perhaps the suggestion/solution would be to add a note  explicitly stating that `put` can't be used in the form at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-method

You might reasonably argue that the text explicitly lists the supported options so we don't need one. I'd accept that. But also I know that people might just assume it an omission in the docs.

**Comment by David263** [comment]
I agree. If the page is going to mention PUT at all, it should say that PUT is not supported (by browsers? by servers?) as an alternative to POST. This could save readers time or misunderstanding.

**Comment by lionralfs** [comment]
> but perhaps the suggestion/solution would be to add a note explicitly stating that `put` can't be used in the form at https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-method

@hamishwillee At that point you might have to add all other HTTP verbs, which adds a bit too much noise in my opinion. I'd rather emphasize on the listed methods being the **only** valid ones.



> I agree. If the page is going to mention PUT at all, it should say that PUT is not supported (by browsers? by servers?) as an alternative to POST. This could save readers time or misunderstanding.

@David263 I can't find any explicit mention of `PUT` on the [form page](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form). `PUT` itself is a valid HTTP method that is widely used (and supported), it's just not a suitable value for the `method` attribute of a `<form>`.

**Comment by David263** [comment]
"|PUT|  itself is a valid HTTP method that is widely used (and supported), it's just not a suitable value for the |method| attribute of a |<form>|."

This is an important point. If neither the MDN Form or PUT article mentions it, then others will likely do as I did: try out PUT and discover that it is implemented as GET. The fact that PUT is a valid method makes it imperative that MDN state that it is not suitable for use in a Form. It might even be useful to say why that is true (PUT is for entering data into a database, not for displaying a page of output).

**Comment by lionralfs** [comment]
> neither the MDN Form or PUT article mentions it

I believe this is what @hamishwillee was trying to explain. In the form article under the `method` attribute you can find the supported options: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/form#attr-method 
The question remains whether it should be further highlighted that these are the *only* two options.

> that it is implemented as GET

You're not seeing an actual `PUT`, the browser is using `GET` as a fallback when using an invalid method in a form.

**Comment by David263** [comment]
My response is that I understand everything that has been said here, and I have offered no objection to the facts.

Yes, the reason I filed this Issue is that I feel strongly that it should be further highlighted that GET and POST are the only two options supported. And it should be explained why and how. The why is that only GET and POST are used to display a page in the browser (the other methods being useful for REST and database queries), and the how is either a reference to a relevant standard or to the software responsible for allowing PUT and silently converting it to GET (either the browsers, the server software, or both).
